storage/innobase/include/gis0type.h

```diff
@@ -66,10 +66,7 @@ typedef std::vector<rtr_rec_t, ut_allocator<rtr_rec_t> >	rtr_rec_vector;
 
 /* Structure for matched records on the leaf page */
 typedef	struct matched_rec {
-	byte*		bufp;		/*!< aligned buffer point */
-	byte		rec_buf[UNIV_PAGE_SIZE_MAX * 2];
-					/*!< buffer used to copy matching rec */
-	buf_block_t	block;		/*!< the shadow buffer block */
+	buf_block_t*	block;		/*!< the shadow buffer block */
```

Comment:
Let's initialize the pointer member to nullptr in  rtr_init_rtr_info when initializing for the first time.
```
rtr_init_rtr_info()
{
...
if (!reinit) {
...
rtr_info->block = nullptr;
}
```

Reply:
[Something is already wrong before this change](https://buildbot.mariadb.org/#/builders/572/builds/7133), observed on 10.11 11c75fc3967d5bf64eed37fdab0ed31af7149756 in the test `innodb_gis.rtree_compress`:
```
==182392==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x560dd9e945cb in rtr_pcur_getnext_from_path(dtuple_t const*, page_cur_mode_t, btr_cur_t*, unsigned long, unsigned long, bool, mtr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/gis/gis0sea.cc:292:3
    #1 0x560ddccf84e0 in rtr_search(dtuple_t const*, btr_latch_mode, btr_pcur_t*, mtr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/gis/gis0sea.cc:1087:8
    #2 0x560ddcd3da49 in row_undo_ins_remove_sec_low(btr_latch_mode, dict_index_t*, dtuple_t*, que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0uins.cc:283:7
    #3 0x560ddcd36569 in row_undo_ins_remove_sec(dict_index_t*, dtuple_t*, que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0uins.cc:353:8
    #4 0x560ddcd36569 in row_undo_ins_remove_sec_rec(undo_node_t*, que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0uins.cc:547:10
    #5 0x560ddcd36569 in row_undo_ins(undo_node_t*, que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0uins.cc:599:9
    #6 0x560ddc705d13 in row_undo(undo_node_t*, que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0undo.cc:399:5
    #7 0x560ddc705d13 in row_undo_step(que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/row/row0undo.cc:440:8
    #8 0x560ddc487d18 in que_thr_step(que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/que/que0que.cc:586:9
    #9 0x560ddc487d18 in que_run_threads_low(que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/que/que0que.cc:644:25
    #10 0x560ddc487d18 in que_run_threads(que_thr_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/que/que0que.cc:664:2
    #11 0x560ddc80d693 in trx_t::rollback_low(trx_savept_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/trx/trx0roll.cc:124:5
    #12 0x560ddc80420b in trx_rollback_for_mysql(trx_t*) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/trx/trx0roll.cc
    #13 0x560ddc09bdb0 in innobase_rollback(handlerton*, THD*, bool) /home/buildbot/amd64-debian-11-msan/build/storage/innobase/handler/ha_innodb.cc:4620:11
    #14 0x560ddb156cea in ha_rollback_trans(THD*, bool) /home/buildbot/amd64-debian-11-msan/build/sql/handler.cc:2236:17
    #15 0x560ddab473fd in trans_rollback(THD*) /home/buildbot/amd64-debian-11-msan/build/sql/transaction.cc:372:8
    #16 0x560dda4b1b3e in mysql_execute_command(THD*, bool) /home/buildbot/amd64-debian-11-msan/build/sql/sql_parse.cc:5784:27
```

Reply:
[The following failure](https://buildbot.mariadb.org/#/builders/572/builds/7234/steps/8/logs/stdio), observed for #3107, looks like something that should be caused by this change:
```
innodb_gis.rtree_compress '16k,innodb'   w7 [ fail ]
        Test ended at 2024-03-07 14:07:27
CURRENT_TEST: innodb_gis.rtree_compress
mysqltest: At line 43: query 'rollback' failed: <Unknown> (2013): Lost connection to server during query
…
2024-03-07 14:07:11 0 [Note] Starting MariaDB 10.6.18-MariaDB-debug-log source revision 63e1c180709dcd9b3234ddd60578bbc0a95b0958 as process 174455
…
mariadbd: /home/buildbot/amd64-debian-11-msan/build/storage/innobase/include/page0cur.inl:61: rec_t *page_cur_get_rec(const page_cur_t *): Assertion `!cur->rec || page_align(cur->rec) == cur->block->page.frame()' failed.
…
#8  0x00005594637fef68 in page_cur_get_rec(page_cur_t const*) ()
#9  0x00005594637f3cee in rtr_page_get_father_block(unsigned short*, mem_block_info_t*, mtr_t*, btr_cur_t*, btr_cur_t*) ()
#10 0x00005594633c14d5 in btr_compress(btr_cur_t*, bool, mtr_t*) ()
#11 0x000055946345654c in btr_cur_compress_if_useful(btr_cur_t*, bool, mtr_t*) ()
#12 0x000055946345e212 in btr_cur_pessimistic_delete(dberr_t*, unsigned long, btr_cur_t*, unsigned long, bool, mtr_t*) ()
#13 0x000055946383412d in row_undo_ins_remove_sec_low(btr_latch_mode, dict_index_t*, dtuple_t*, que_thr_t*) ()
#14 0x000055946382c96b in row_undo_ins(undo_node_t*, que_thr_t*) ()
#15 0x0000559463213d00 in row_undo_step(que_thr_t*) ()
#16 0x0000559462fa28d9 in que_run_threads(que_thr_t*) ()
#17 0x0000559463325c9f in trx_t::rollback_low(trx_savept_t*) ()
#18 0x000055946331c801 in trx_rollback_for_mysql(trx_t*) ()
#19 0x0000559462bdb3d1 in innobase_rollback(handlerton*, THD*, bool) ()
#20 0x0000559461be5a9b in ha_rollback_trans(THD*, bool) ()
```

Reply:
The failure of the test `innodb_gis.rtree_compress` in `rollback` appears to be caused by a race condition between `buf_flush_page_cleaner()` freeing a page 5:0x53 and `rtr_pcur_getnext_from_path()`. Apparently, that function is not holding a buffer-fix or anything to prevent page eviction. Thanks to `buf_LRU_block_remove_hashed()` invoking `MEM_UNDEFINED(page, srv_page_size)`, this will lead to `__msan_warning_with_origin_noreturn()` being invoked when a debug assertion is trying to access the field `PAGE_LEVEL` in the now-freed page. I will try to fix that unrelated-to-this-fix race condition anyway, so that I can reproduce and debug the second one.

Reply:
The debug assertion is misplaced, and I think it is simplest to remove it. Additionally, it is wrong to use `RW_NO_LATCH` without checking that the block was actually successfully loaded with the expected page identifier. It is better to initially acquire `RW_S_LATCH` or `RW_X_LATCH`, and downgrade it to a buffer-fix before invoking `rtr_latch_leaves()`. After I did that, I was able to reproduce the assertion failure.

Reply:
The assertion failure looks unrelated to this change as well. We had stored the cursor position on page 4 in the tablespace 5 (the spatial index root page):
```
#0  0x0000564708682a4e in page_cur_position (rec=rec@entry=0x7fde8889c0a8 "", block=block@entry=0x7fde88801900, cur=cur@entry=0x70f000007cc8) at /mariadb/10.6/storage/innobase/include/page0cur.inl:137
#1  0x00005647086826ad in rtr_non_leaf_insert_stack_push (index=index@entry=0x715000091320, path=0x7040000170d8, level=level@entry=1, child_no=child_no@entry=8, block=block@entry=0x7fde88801900, 
    rec=rec@entry=0x7fde8889c0a8 "", mbr_inc=<optimized out>) at /mariadb/10.6/storage/innobase/gis/gis0sea.cc:1851
#2  0x00005647086742df in rtr_cur_search_with_match (block=block@entry=0x7fde88801900, index=index@entry=0x715000091320, tuple=tuple@entry=0x719000041fa0, mode=PAGE_CUR_WITHIN, 
    mode@entry=PAGE_CUR_RTREE_LOCATE, cursor=cursor@entry=0x7fde8bd7d570, rtr_info=0x711000085c18) at /mariadb/10.6/storage/innobase/gis/gis0sea.cc:2178
#3  0x000056470868962d in rtr_search_to_nth_level (level=level@entry=0, tuple=tuple@entry=0x719000041fa0, mode=mode@entry=PAGE_CUR_RTREE_LOCATE, latch_mode=<optimized out>, latch_mode@entry=4618, 
    cur=cur@entry=0x7fde8bd7d570, mtr=mtr@entry=0x7fde8bd7d748) at /mariadb/10.6/storage/innobase/gis/gis0sea.cc:799
#4  0x000056470867859f in rtr_search_leaf (cur=0x7fde8bd7d570, tuple=0x719000041fa0, latch_mode=4618, mtr=0x7fde8bd7d748, mode=PAGE_CUR_RTREE_LOCATE) at /mariadb/10.6/storage/innobase/gis/gis0sea.cc:990
#5  rtr_search (tuple=tuple@entry=0x719000041fa0, latch_mode=4618, cursor=cursor@entry=0x7fde8bd7d570, mtr=mtr@entry=0x7fde8bd7d748) at /mariadb/10.6/storage/innobase/gis/gis0sea.cc:1063
#6  0x0000564708b50b35 in row_undo_ins_remove_sec_low (mode=0, mode@entry=BTR_PURGE_TREE, index=index@entry=0x715000091320, entry=entry@entry=0x719000041fa0, thr=thr@entry=0x716000041e58)
    at /mariadb/10.6/storage/innobase/row/row0uins.cc:283
#7  0x0000564708b4bad2 in row_undo_ins_remove_sec (index=0x715000091320, entry=0x719000041fa0, thr=0x716000041e58) at /mariadb/10.6/storage/innobase/row/row0uins.cc:362
#8  row_undo_ins_remove_sec_rec (node=0x71a0000402a0, thr=0x716000041e58) at /mariadb/10.6/storage/innobase/row/row0uins.cc:547
```
At the time of the assertion failure, we are positioning the cursor on page 4 in that tablespace, but meanwhile the page had been evicted and reloaded into the buffer pool. We are reading a `page_cur_t::rec` that is pointing to a stale copy of the page descriptor.

Reply:
That assertion fails due to an incorrect refactoring in #3107. Previously, we were reading `page_cur->block.page.frame`; now we are invoking `page_cur_get_page(page_cursor)` before `page_cur->rec` had been initialized by `page_cur_search_with_match()`. If we determine `page_ssn` after that call succeeded, everything seems to be fine.

Reply:
>Let's initialize the pointer member to `nullptr` in `rtr_init_rtr_info` when initializing for the first time.

In fact, we were already kind of doing that:
```c++
	if (!reinit) {
		/* Reset all members. */
		rtr_info->path = NULL;
		rtr_info->parent_path = NULL;
		rtr_info->matches = NULL;
```
Because the field `matches` is a null pointer, it can’t point to something where we would have forgotten to initialize the data member `block=nullptr`.

Reply:
I filed #3185 for that race condition that I mentioned. It is unclear if it would affect older versions than 10.6.